mesonbuild/modules/unstable_external_project.py

```diff
@@ -0,0 +1,250 @@
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os, subprocess, shlex
+
+from . import ExtensionModule, ModuleReturnValue
+from .. import mlog, build
+from ..mesonlib import (MesonException, Popen_safe, MachineChoice,
+                       get_variable_regex, do_replacement)
+from ..interpreterbase import InterpreterObject, InterpreterException, FeatureNew
+from ..interpreterbase import stringArgs, permittedKwargs
+from ..interpreter import DependencyHolder, InstallDir
+from ..compilers.compilers import cflags_mapping, cexe_mapping
+from ..dependencies.base import InternalDependency, PkgConfigDependency
+
+class ExternalProject(InterpreterObject):
+    def __init__(self, interpreter, subdir, project_version, subproject, environment, build_machine, host_machine,
+                 configure_command, configure_options, cross_configure_options, env, verbose):
+        InterpreterObject.__init__(self)
+        self.methods.update({'dependency': self.dependency_method,
+                             })
+
+        self.interpreter = interpreter
+        self.subdir = subdir
+        self.project_version = project_version
+        self.subproject = subproject
+        self.env = environment
+        self.build_machine = build_machine
+        self.host_machine = host_machine
+        self.configure_command = configure_command
+        self.configure_options = configure_options
+        self.cross_configure_options = cross_configure_options
+        self.verbose = verbose
+        self.user_env = env
+
+        self.name = os.path.basename(self.subdir)
+        self.src_dir = os.path.join(self.env.get_source_dir(), self.subdir)
+        self.build_dir = os.path.join(self.env.get_build_dir(), self.subdir, 'build')
+        self.install_dir = os.path.join(self.env.get_build_dir(), self.subdir, 'dist')
+        self.prefix = self.env.coredata.get_builtin_option('prefix')
+        self.libdir = self.env.coredata.get_builtin_option('libdir')
+        self.includedir = self.env.coredata.get_builtin_option('includedir')
+
+        # On Windows if the prefix is "c:/foo" and DESTDIR is "c:/bar", `make`
+        # will install files into "c:/bar/c:/foo" which is an invalid path.
+        # Work around that issue by removing the drive from prefix.
+        self.prefix = os.path.splitdrive(self.prefix)[1]
+
+        # self.prefix is an absolute path, so we cannot use it in
+        # os.path.join(something, self.prefix).
+        self.rel_prefix = self.prefix[1:]
+
+        self.make = self.interpreter.find_program_impl(os.environ.get('MAKE', 'make'))
```

Comment:
Shouldn't `os.environ` be avoided?

Reply:
Is there another way to get $MAKE from env? I got told that it's commonly passed by env, like PKG_CONFIG, etc.

Reply:
If I remember correctly, there was a function that automatically gets the correct value from the env and machine files. Not sure how exactly it worked and where it is.

Reply:
Ok you're right, find_program_impl('make') will now return the command from MAKE env, or machine file.